web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts

```diff
@@ -48,23 +74,121 @@ export default function useDocumentTerminal(doc: Doc) {
     };
   }, [state]);
 
-  // TODO: Move this within the call to initState in useAsync.
-  useEffect(() => {
-    if (state.status === 'error') {
-      ctx.notificationsService.notifyError({
-        title: 'Could not open a terminal',
-        description: state.statusText,
-      });
+  return state;
+}
+
+async function startTerminalSession(
+  ctx: IAppContext,
+  logger: Logger,
+  documentsService: DocumentsService,
+  doc: types.DocumentTerminal
+) {
+  if (doc.kind === 'doc.terminal_tsh_node' && !('serverId' in doc)) {
```

Comment:
> I'd try to convert this entire condition to a user-defined type guard: `isDocumentTshNodeWithLoginHost()`.

Hmm okay. I thought that custom type guards have the potential to make your code unsafe, as in TypeScript just trusts whatever is inside them without actually performing any typechecking itself. But I re-read the docs and I think it's not the case, so I'll try to implement it here.

Reply:
It looks like it actually has the potential to make code unsound:

```typescript
export function isDocumentTshNodeWithLoginHost(
  doc: Document
): doc is DocumentTshNodeWithLoginHost {
  return doc.kind === 'doc.terminal_tsh_node' && 'serverId' in doc;
}
```

In TypeScript's eyes, this is a valid function even though `DocumentTshNodeWithLoginHost` does not have a serverId field.

Thoughts on this? I'd prefer to leave custom type guards for cases where it's impossible to guarantee type safety otherwise. A good example would be functions from `uri.ts` such as `isLeafCluster` where the type guard could narrow the type.

Reply:
Eh :(

My main problem is that `'serverId' in doc` can mean two things:
- it is used to narrow down the type (we had two types -> now we have one)
- there is only one type and we only want to check if the field is present in it (because it is optional)

I think a type guard is not that bad in this case.
But if we don't want to go with it, maybe we could do at least this:
```ts
const isDocumentTshNodeWithLoginHost = (doc.kind === 'doc.terminal_tsh_node' && !('serverId' in doc))`
if (isDocumentTshNodeWithLoginHost) {}
```

Reply:
> My main problem is that `'serverId' in doc` can mean two things

Agree that `'field' in doc` might be ambiguous, though in the second case when the field is optional you typically check the existence of the value under the field with `if (doc.field)`. Unlike during type narrowing, TypeScript won't complain about `doc` not having `field`, which is the sole reason why we use `'field' in doc` for type narrowing rather than `doc.field`.

Anyway, I think I'll actually add a function called `isDocumentTshNodeWithLoginHost` to documentsService/types.ts, just without the type guard. TypeScript is still going to narrow down the type when executing a branch protected with this function.

I'm not sure if this is what you had in mind as your usage of `const isDocumentTshNodeWithLoginHost` suggests to me that you wanted to keep it local to `useDocumentTerminal`.

Reply:
I'm going to add `isDocumentTshNodeWithLoginHost` in the next PR as well since it includes other changes related to types.

Reply:
Type narrowing actually stops working if you move it to a separate function which just returns bool. 😵‍💫

Alright then, I'll add a type guard.

Reply:
>Type narrowing actually stops working if you move it to a separate function which just returns bool. 

Yeah, it works only when you return the value you checked from the function :( That's I though about a local variable, but the type guard should be the most convenient thing to use.